### PR TITLE
dialects: (arith) make parsing robust with respect to block ordering

### PR DESCRIPTION
**Pull request by tobiasgrosser**
The function Parse.parse_operand only works, if we are certain that all operands are defined before they are used. As arith can be used in arbitrary control flow graphs, we need to use the robust pattern of Parser.parse_unresolved_operand + Parser.resolve_operands.

Going forward we will remove remaining uses of parse_operand and eventually rename parse_unresolved_operand to parse_operand.

### CODE REVIEW

**Review by math-fehr — APPROVED** [review]
:+1:

File: tests/xdsl_opt/test_xdsl_opt.py [hunk 1]
```diff
@@ -156,7 +157,7 @@ def test_diagnostic_exception():
 
     opt = xDSLOptMain(args=[filename_in])
 
-    with pytest.raises(DiagnosticException):
+    with pytest.raises(ParseError):
```
Comment (tobiasgrosser) [inline tests/xdsl_opt/test_xdsl_opt.py:160]: This change is needed, as the Parser does now the type-checking and now throws an error: 
> operand is used with type f32, but has been previously used or defined with type i32

Is there a specific reason that we do type-checking in the parser. I feel this is somehow the wrong place to do type checking as it e.g. prevents further type errors at other locations to be emitted.

Comment (math-fehr) [reply tests/xdsl_opt/test_xdsl_opt.py:159]: Let me check if we should instead change the program that is being parsed there.

The reason this is checked at parse time and not after, is because the program cannot even be constructed.
If you refer to a value that is expected to have type `i32`, and then you parse that value definition which has a different type, let's say `i64`, then you have a fundamental inconsistency in your program.
It's not even that it is ill-typed, it is that a value has two different types, and thus we should signal it directly.

**Comment by math-fehr** [comment]
I quickly fixed the test so it still check the same thing as before, which is that a unverified program will return a diagnostic error.

File: tests/xdsl_opt/test_xdsl_opt.py (hunk 1 above)
Comment (tobiasgrosser) [reply tests/xdsl_opt/test_xdsl_opt.py]: I am confused. In which way would a type error be `fundamentally inconsistent`?

Comment (math-fehr) [reply tests/xdsl_opt/test_xdsl_opt.py]: What I mean is that you cannot even create an IR that correspond to that program, because you have a value that has two different types.

Also, note that this is different than having an operation not verifying. This is handled later.

Comment (math-fehr) [reply tests/xdsl_opt/test_xdsl_opt.py]: Also, this is exactly how it is done in MLIR, just in case.

File: tests/xdsl_opt/unverified_program.mlir [hunk 2]
```diff
@@ -1,5 +1,5 @@
 builtin.module {
     %a = "arith.constant"() {value = 1 : i32} : () -> i32
     %b = "arith.constant"() {value = 1.0 : f32} : () -> f32
-    %c = arith.addf %a, %b : f32
+    %c = "arith.addf"(%a, %b) : (i32, f32) -> f32
```
Comment (superlopuh) [inline tests/xdsl_opt/unverified_program.mlir:4]: why this change?

**Review by superlopuh — APPROVED** [review]
thank you

File: tests/xdsl_opt/unverified_program.mlir (hunk 2 above)
Comment (superlopuh) [reply tests/xdsl_opt/unverified_program.mlir:4]: Ah I missed this discussion in the main thread of this PR

File: tests/filecheck/dialects/arith/arith_cfg.mlir [hunk 3]
```diff
@@ -0,0 +1,32 @@
+// RUN: xdsl-opt %s
```
Comment (webmiche) [inline tests/filecheck/dialects/arith/arith_cfg.mlir:1]: Could we use the test dialcet here?

Also is this really a property of arith or just a property of the ir-structure in general? If so, we could propably replace everything with the test dialect and move this to a new folder "ir-structure" or something.

**Review by webmiche — COMMENTED** [review]
LGTM, modulo using the test dialect.

File: tests/filecheck/dialects/arith/arith_cfg.mlir (hunk 3 above)
Comment (tobiasgrosser) [reply tests/filecheck/dialects/arith/arith_cfg.mlir:1]: No, we can unfortunately not. This test case tests the custom parsing of the `arith.add` operation, so we need at least arith to stick around. Or are you suggesting I replace the func.func and cf.br operation with the test dialect? I guess that could be possible.

Comment (webmiche) [reply tests/filecheck/dialects/arith/arith_cfg.mlir:1]: Yes, I think the func and cf dialects could certainly go.

Ah yes, I see the changes happen within arith, so that one can't go :+1:

Comment (tobiasgrosser) [reply tests/filecheck/dialects/arith/arith_cfg.mlir:1]: The test.termop does not allow `> 0` successors. Are you suggesting we expand termop to allow this?

Comment (webmiche) [reply tests/filecheck/dialects/arith/arith_cfg.mlir:1]: Yes, that would be needed too, I guess. IMO, you can do this in this PR, I don't mind. It feels like a rather small change...

Comment (tobiasgrosser) [reply tests/filecheck/dialects/arith/arith_cfg.mlir:1]: I tried this, but the build breaks as I cannot get var_successor_def to work.

**Review by AntonLydike — APPROVED** [review]
LGTM, modulo some pyright issues it seems?

File: tests/filecheck/dialects/arith/arith_cfg.mlir (hunk 3 above)
Comment (tobiasgrosser) [reply tests/filecheck/dialects/arith/arith_cfg.mlir:1]: @webmiche, can you see if you can get the test dialect fixed here?

File: xdsl/dialects/test.py [hunk 4]
```diff
@@ -34,7 +36,7 @@ class TestOp(IRDLOperation):
 
 
 @irdl_op_definition
-class TestTermOp(TestOp):
+class TestTermOp(IRDLOperation):
```
Comment (webmiche) [inline xdsl/dialects/test.py:39]: @math-fehr it seems that this was broken before. This would only collect the constraints of the parent op for verification and not the ones specified within this class. Do we want it to be that way? Or did I just bandage over an IRDL bug?

File: xdsl/dialects/test.py [hunk 5]
```diff
@@ -46,6 +48,10 @@ class TestTermOp(TestOp):
 
     name = "test.termop"
 
+    res: VarOpResult = var_result_def()
```
Comment (webmiche) [inline xdsl/dialects/test.py:51]: Does it make sense for a terminator to have results?
This is something we used in some of the tests. That's why my earlier try failed ([CI run](https://github.com/xdslproject/xdsl/actions/runs/5611853160/jobs/10268770422))

File: tests/filecheck/dialects/arith/arith_cfg.mlir (hunk 3 above)
Comment (webmiche) [reply tests/filecheck/dialects/arith/arith_cfg.mlir:1]: Fixed it. Let's see what Mathieu thinks about my fixes and then we can merge, I guess.

File: xdsl/dialects/test.py (hunk 5 above)
Comment (webmiche) [reply xdsl/dialects/test.py:51]: My one question was also about SSAValue scoping which I was not 100% sure about anymore. If SSAValues are block-scoped, this does not make sense as the terminator is the last operation in the block. If they are region-scoped (which from a quick google search, they seem to be in MLIR), I guess you could build IR that uses it.

And then the follow up question is: Will there ever be a use for it? Who would define a terminator that defines an SSAValue?

I guess we can leave it as-is. Better test more behavior than what people actually use. But, it was something that made me raise my eyebrow, so I decided to comment on it :sweat_smile:

Comment (compor) [reply xdsl/dialects/test.py:51]: > I guess you could build IR that uses it.

Yeah, true, I missed that point.
But my doubts remain on the utility of this, and when in Rome, I'd rather do what MLIR does (or does not do in this case :-)

File: xdsl/dialects/test.py (hunk 4 above)
Comment (math-fehr) [reply xdsl/dialects/test.py:39]: Seems like a :bug: to me yes!

File: xdsl/dialects/test.py (hunk 5 above)
Comment (math-fehr) [reply xdsl/dialects/test.py:51]: Technically, a terminator can have regions, and even results.
Should they? I don't think so, but *maybe* someone one day will have a use case for it, so we can keep it as MLIR does.

**Review by math-fehr — APPROVED** [review]
:+1:

File: xdsl/dialects/test.py (hunk 5 above)
Comment (webmiche) [reply xdsl/dialects/test.py:51]: I added regions as well. Let's keep it like this...

File: xdsl/dialects/test.py (hunk 4 above)
Comment (webmiche) [reply xdsl/dialects/test.py:39]: opened #1312

**Comment by webmiche** [comment]
I think we can merge after a rebase.
